### DISCUSSION TITLE
faiss vs milvus search

### DISCUSSION THREAD

**Discussion by Yufei447** [milvus-io/milvus#25938] [Q&A and General discussion]
Now I want to compare the searching result returning by FAISS and milvus. I want to build the exactly same index and use same parameter to search. 

Using FAISS, I create index like this:
quantizer = faiss.IndexFlatIP(DIMENSION)
index = faiss.IndexIVFFlat(quantizer, DIMENSION, NLIST, faiss.METRIC_INNER_PRODUCT)

Before search, I normalize the vector to be search:
faiss.normalize_L2(vectors)

Using Milvus, I create index like this:
index = {"index_type":"IVF_FLAT", "metric_type": "IP", "params": {"nlist": NLIST}}

The search params I specify is like this:
search_params = {"metric_type": "IP", "params": {"nprobe": NPROBE}}

I have 3 questions:
1. FAISS can build quantizer, can milvus build quantizer like FAISS? If not, is there any performance difference?
2. FAISS has function normalize_L2() to normalize vectors, do milvus has function which can achieve same function?
3. Is the index I build using FAISS the same as I build with Milvus? If so, why is there such a big difference between the result returned by Milvus and the result returned by FAISS?
4. For the result returning by milvus, there is a field called "distance". But I specify the search metric to be IP(inner product), why the field is distance not similarity?

**Comment by yhmo** [comment]
1. Milvus supports these index types: IVF_FLAT, IVF_PQ, IVF_SQ8, HNSW.   In these types, IVF_PQ and IVF_SQ8 is quantizer index.
2. Milvus doesn't normalize vectors.
3. I think IVF_FLAT is similar to FAISS, some specializes to the IVF_PQ and IVF_SQ8. Milvus manages data by segments, and build an independent index for each segment, search segments one by one, each index generate a result, and combine these results into a final result. FAISS simply builds an index for the entire dataset(if the dataset is constant).
4. "distance" is just a name, users know what the value is.

**Comment by xiaofan-luan** [comment]
> 

2. you can use faiss normalize_L2 for milvus as well.
3. Milvus should return more accurate result with same index parameter, but it shouldn't be too much different

**Comment by Yufei447** [comment]
I defined the searching metric of milvus as “IP”, and it returned results greater than 1. But shouldn't the inner product similarity be between -1 and 1? Why are all my search results returning distances greater than 1?

**Reply by yhmo** [reply]
Because the vectors are not normalized.

Normalize the vectors on the client side:
```
# these are my vectors
vectors = generate_my_vectors()

# normalize the vectors
normalized_vectors = normalize_method(vectors)

# insert normalized vectors into milvus
collection = Collection("xxx")
collection.insert(data = [
    ......
    normailized_vectors
])
```

**Reply by Yufei447** [reply]
I already called faiss.normalize_L2(vectors_to_search) before searching

**Reply by yhmo** [reply]
Both the vectors in the collection and the **vectors_to_searh** must be normalized.

**Reply by Yufei447** [reply]
ic. I'll try it another time. Thanks so much
